### ISSUE TITLE
z.enum to allow non-string options

### ISSUE THREAD

**Issue by IlyaSemenov** [colinhacks/zod#2317]
I recently opened the issue #2312 where I elaborated that the typings of `z.enum` were too restrictive as they didn't allow arbitrary list of strings.

This is the extension of the above problem which goes not only about typings but also about implementation. _(I am opening a separate issue as I realise this is all breaking changes and maintainers tend to disregard them... so I give these separate issues more chance to survive :)_

### Problem

Imagine I have a schema that validates a value to be one of the predefined integers. Like when a user chooses the length of the secret key to be either 512, 1024 or 2048 bits (just random example). Naturally, I would imagine there was a Zod schema for that... like `z.enum([512, 1024, 2048])`? Unfortunately, that doesn't work, as `z.enum` only accepts string values.

### Current workaround

One can use `z.union([z.literal(512), z.literal(1024), z.literal(2048)]` but that's cumbersome (and also arguably slow as it validates each sub-schema independently).

### Suggested approach

What if `z.enum` simply allowed any list of anything as valid options?

Like, what if `z.enum([1, 5, "testing", true])` simply worked and validated the value to be one of the options (with `options.includes(value)` as the implementation)? The resulting type could be derived from the options with [`ValueOf<Options>`](https://github.com/ts-essentials/ts-essentials/tree/afccd3550cc1c03b87d1df4bd590b1e1caa0c930/lib/value-of) recipe.

**Comment by igalklebanov** [comment]
Hey 👋

> `z.enum` is a Zod-native way to declare a schema with a fixed set of allowable string values.
> To get autocompletion with a Zod enum, use the `.enum` property of your schema:
> ~ README.md

Zod enums are for string values known at compile-time, so they could provide autocompletion.

What happens if you pass `z.enum([5, '5'])`, what should the type and runtime value of `parsed.enum['5']` be?

**Comment by IlyaSemenov** [comment]
@igalklebanov I agree, that is valid point. I didn't consider `schema.enum` and it will indeed break. (I must have disregarded it as I saw no point in it for my scenarios, but I realise that in some circumstances it makes sense to have it).

However, I don't think `z.union` should be changed. It is by its nature an aggregating schema builder and should do that. Instead, I believe it's better to introduce `z.options` which will be basically `z.enum` without the `enum` auto-completion magic but accepting anything as the list of options (not even necessarily primitives).
